### PR TITLE
Adding ASR

### LINKED ISSUE #85: [FEATURE-REQUEST] Adding Automatic Speech Recognition
**Is your feature request related to a problem? Please describe.**
Add the ASR task.

### PR DESCRIPTION
**Pull request by AJDERS**
This PR introduces:

- Automatic Speech Recognition task.
- A new field in `task_config` called `architectures` which contains valid architectures for the task, this is relevant for tasks such as ASR which has multiple valid architectures, e.g. Whisper and Wav2Vec2ForCTC.
- Most places in `task.py` where there previously was passed a `tokenizer` there is now passed `tokenizer_or_processor`, again to accommodate ASR. 
- Updates to tests pertaining to tokenizer or checking of supertask.

This closes #85.

### CODE REVIEW

File: src/aiai_eval/utils.py [hunk 1]
```diff
@@ -203,22 +207,39 @@ def check_supertask(architectures: Sequence[str], supertask: str):
...
-    # architectures
-    supertask_is_an_architecture = any(
-        kebab_to_pascal(supertask) in architecture for architecture in architectures
-    )
+    # Create list which contains the supertask if it exists among the available architectures
+    supertask_which_is_architectures = [
+        supertask
+        for architecture in architectures
+        if kebab_to_pascal(supertask) in architecture
+    ]
```
Comment (saattrupdan) [inline src/aiai_eval/utils.py:225]: This list will always have exactly 0 or 1 element, which seems a bit silly. Why not use the previous Boolean value?

File: src/aiai_eval/utils.py [hunk 2]
```diff
@@ -203,22 +207,39 @@ def check_supertask(architectures: Sequence[str], supertask: str):
             The supertask associated to a task written in kebab-case, e.g.,
             text-classification.
 
+    Returns:
+        tuple of list of str:
+            The architectures that are valid for the supertask, and the provided model architectures that
+            which are allowed for the supertask.
```
Comment (saattrupdan) [inline src/aiai_eval/utils.py:213]: This is quite confusion. What's the difference between "valid" and "allowed"?

File: tests/conftest.py [hunk 3]
```diff
@@ -55,6 +55,7 @@ def model_configs(evaluation_config, task_config):
         ],
         "question-answering": ["deepset/minilm-uncased-squad2"],
         "offensive-text-classification": ["DaNLP/da-electra-hatespeech-detection"],
+        "automatic-speech-recognition": ["openai/whisper-small"],
```
Comment (saattrupdan) [inline tests/conftest.py:58]: Maybe use tiny, to make tests faster?

File: tests/test_utils.py [hunk 4]
```diff
@@ -129,38 +129,47 @@ def test_clear_memory():
 
 
 @pytest.mark.parametrize(
-    argnames="architectures,supertask,raises_error",
+    argnames="architectures,supertask,allowed_architectures,raises_error",
     argvalues=[
         (
             ["TokenClassification", "SequenceClassification"],
             "token-classification",
+            "token-classification",
```
Comment (saattrupdan) [inline tests/test_utils.py:137]: Shouldn't `allowed_architectures` be a sequence?

File: pyproject.toml [hunk 5]
```diff
@@ -26,6 +25,13 @@ fsspec = "^2022.7.1"
 termcolor = "^1.1.0"
 gradio = "^3.1.7"
 colorama = ">=0.4.5,<1.0.0"
+jiwer = "^2.5.1"
+transformers = "^4.23.0"
+soundfile = "^0.11.0"
+torchaudio = "^0.12.1"
+evaluate = "^0.3.0"
+pyctcdecode = "^0.4.0"
+pypi-kenlm = "^0.1.20220713"
```
Comment (saattrupdan) [inline pyproject.toml:34]: We should maybe consider making these dependencies optional.

**Review by saattrupdan — CHANGES_REQUESTED** [review]

File: src/aiai_eval/automatic_speech_recognition.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            input_features = [
+                {
+                    "input_values": self.processor(
+                        feature["input_values"]["array"],
+                        sampling_rate=sampling_rate,
+                        padding=self.padding,
+                    ).input_values[0]
+                }
+                for feature in features
+            ]
```
[the remark this answers is not in the document]
Comment (AJDERS) [reply src/aiai_eval/automatic_speech_recognition.py:83]: Not the only difference, one has `input_values[0]` the other has `input_features[0]`.

File: src/aiai_eval/automatic_speech_recognition.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+        # Create batch from input_features
+        batch = self.processor.feature_extractor.pad(
+            input_features,
+            padding=self.padding,
+            return_tensors="pt",
+        )
```
[the remark this answers is not in the document]
Comment (AJDERS) [reply src/aiai_eval/automatic_speech_recognition.py:90]: True!

File: src/aiai_eval/task.py [hunk 8]
```diff
@@ -255,7 +261,13 @@ def _evaluate_single_iteration(
             torch.cuda.manual_seed_all(703 + idx)
 
             # Define batch size, which depends on whether we are testing or not
-            batch_size = 2 if self.evaluation_config.testing else 32
+            batch_size = (
+                2
+                if self.evaluation_config.testing
+                else 4
+                if self.task_config.name == "automatic-speech-recognition"
```
[the remark this answers is not in the document]
Comment (AJDERS) [reply src/aiai_eval/task.py:268]: Yup! Ill open an issue.

File: src/aiai_eval/task.py [hunk 9]
```diff
@@ -516,9 +574,9 @@ def _get_model_predictions(
                             ],
                             dim=-1,
                         )
-
-                    # Otherwise, we raise an error
-                    else:
+                    # In case of ASR, we recieve a tensor, and we if we do not, then
+                    # we throw an error.
+                    elif not isinstance(model_predictions, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (AJDERS) [reply src/aiai_eval/task.py:579]: `model_predictions = model.generate` outputs `torch.Tensor`'s which would give an error with the previous code. Now if the output does not have `logits` or `start_logits`/`end_logits`, we check if it is even a tensor, and if not, well we throw an error.

Would a more descriptive comment help? Or are you thinking something else? :)

File: src/aiai_eval/task.py [hunk 10]
```diff
@@ -540,7 +598,7 @@ def _get_model_predictions(
                     raise UnsupportedModelType(model_type=model_type)
 
                 # Move the predictions back to the CPU and convert it to a NumPy array
-                model_predictions = model_predictions.cpu().numpy().tolist()
+                model_predictions = model_predictions.cpu().numpy()
 
                 # Collect predictions
                 all_predictions.extend(model_predictions)
```
[the remark this answers is not in the document]
Comment (AJDERS) [reply src/aiai_eval/task.py:604]: Yup :)

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (AJDERS) [reply pyproject.toml:34]: Yes that might be a good idea.

File: pyproject.toml [hunk 11]
```diff
@@ -26,6 +24,15 @@ fsspec = "^2022.7.1"
...
 colorama = ">=0.4.5,<1.0.0"
+jiwer = "^2.5.1"
+transformers = "^4.23.0"
+soundfile = "^0.11.0"
+evaluate = "^0.3.0"
+pyctcdecode = "^0.4.0"
+pypi-kenlm = "^0.1.20220713"
+librosa = "^0.9.2"
+torch = "^1.13.1"
+torchaudio = "^0.13.1"
```
Comment (saattrupdan) [inline pyproject.toml:35]: The behaviour of "^" is a bit odd for some reason. For instance, "^1.2.3" means ">=1.2.3, <2.0.0", which is correct, but "^0.1.2" means ">=0.1.2, <0.2.0", so suddenly we're not supporting the next minor version. We can fix this by making these "zero-major versions" more explicit:

```suggestion
soundfile = ">=0.11.0,<1.0.0"
evaluate = ">=0.3.0,<1.0.0"
pyctcdecode = ">=0.4.0,<1.0.0"
pypi-kenlm = ">=0.1.20220713,<1.0.0"
librosa = ">=0.9.2,<1.0.0"
torch = "^1.13.1"
torchaudio = ">=0.13.1,<1.0.0"
```

File: src/aiai_eval/config.py [hunk 12]
```diff
@@ -71,6 +71,8 @@ class TaskConfig:
             to None.
         supertask (str):
             The supertask of the task, describing the overall type of task.
+        modality str:
```
Comment (saattrupdan) [inline src/aiai_eval/config.py:74]:
```suggestion
        modality (str):
```

File: src/aiai_eval/config.py [hunk 13]
```diff
@@ -71,6 +71,8 @@ class TaskConfig:
...
+            The modality of the input data.
```
Comment (saattrupdan) [inline src/aiai_eval/config.py:75]: Maybe mention what the choices are?

File: src/aiai_eval/config.py [hunk 14]
```diff
@@ -90,17 +92,21 @@ class TaskConfig:
             The number of labels in the dataset.
         label_synonyms (list of list of str):
             The synonyms of all the labels, including the main label.
+        architectures (None or list of str):
+            The architectures that can be used to solve the task. Defaults to None.
```
Comment (saattrupdan) [inline src/aiai_eval/config.py:96]:
```suggestion
            The architectures that can be used to solve the task. If None then
            it defaults to the list containing only the name of the supertaks. Defaults 
            to None.
```

File: src/aiai_eval/hf_hub_utils.py [hunk 15]
```diff
@@ -128,6 +151,18 @@ def load_model_from_hf_hub(
         **params,
     )
 
+    # try to load a processor from the model id, if it does not exist, then set
```
Comment (saattrupdan) [inline src/aiai_eval/hf_hub_utils.py:154]:
```suggestion
    # Try to load a processor from the model id, if it does not exist, then set
```

File: src/aiai_eval/metric_configs.py [hunk 16]
```diff
@@ -75,3 +75,11 @@
     results_key="power",
     postprocessing_fn=lambda raw_score: f"{raw_score:.4f}",
 )
+
+WER = MetricConfig(
+    name="word_error_rate",
+    pretty_name="Word error rate",
+    huggingface_id="wer",
+    results_key="wer",
+    postprocessing_fn=lambda raw_score: str(raw_score),
```
Comment (saattrupdan) [inline src/aiai_eval/metric_configs.py:84]: Why not truncate the number of decimals here, like the others?
```suggestion
    postprocessing_fn=lambda raw_score: f"{raw_score:.4f}",
```

File: src/aiai_eval/named_entity_recognition.py [hunk 17]
```diff
@@ -123,8 +124,10 @@ def _check_if_model_is_trained_for_task(self, model_predictions: list) -> bool:
 
         sample_preds = model_predictions[0]
         has_sequence_elements = len(sample_preds[0]) > 0
-        leaves_are_floats = isinstance(sample_preds[0][0], float)
         elements_are_strings = isinstance(sample_preds[0], str)
+        leaves_are_floats = False
+        if not elements_are_strings:
+            leaves_are_floats = sample_preds[0][0].dtype.kind == "f"
```
Comment (saattrupdan) [inline src/aiai_eval/named_entity_recognition.py:130]: A slight simplification. Mostly for style points, I don't mind if you leave it as is.

```suggestion
        leaves_are_floats = not elements_are_strings and sample_preds[0][0].dtype.kind == "f"
```

File: src/aiai_eval/task.py [hunk 18]
```diff
@@ -463,7 +488,17 @@ def _get_model_predictions(
         if framework == Framework.PYTORCH:
 
             # Load the data collator
-            data_collator = self._load_data_collator(tokenizer)
+            # If the processor is not a tokenizer we assume its a processor
```
Comment (saattrupdan) [inline src/aiai_eval/task.py:491]:
```suggestion
            # If the processor is not a tokenizer we assume it's a processor, and
```

File: tests/conftest.py [hunk 19]
```diff
@@ -54,7 +54,10 @@ def model_configs(evaluation_config, task_config):
             "spacy/da_core_news_sm",
         ],
         "question-answering": ["deepset/minilm-uncased-squad2"],
-        "offensive-text-classification": ["DaNLP/da-electra-hatespeech-detection"],
+        "offensive-text-classification": [
+            "alexandrainst/da-electra-hatespeech-detection"
```
Comment (saattrupdan) [inline tests/conftest.py:58]: Changed the model name again, sorry. All models in our repo now follow the format "<language>-<task>-<size>".

```suggestion
            "alexandrainst/da-hatespeech-detection-small"
```

File: tests/conftest.py [hunk 20]
```diff
@@ -93,12 +96,13 @@ def model_total_scores(model_configs):
             qa_f1=75.0,
             qa_f1_se=49.0,
         ),
-        "DaNLP/da-electra-hatespeech-detection": dict(
+        "alexandrainst/da-electra-hatespeech-detection": dict(
```
Comment (saattrupdan) [inline tests/conftest.py:99]: Same as above.

```suggestion
        "alexandrainst/da-hatespeech-detection-small": dict(
```

File: src/aiai_eval/config.py (hunk 13 above)
Comment (saattrupdan) [reply src/aiai_eval/config.py:75]: Consider making the modalities into an enum, just like `Framework`, to make it clearer.

**Review by saattrupdan — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 21]
```diff
@@ -11,8 +11,6 @@ license = "MIT"
 
 [tool.poetry.dependencies]
 python = ">=3.8,<3.11"
-torch = "^1.12.0"
-transformers = "^4.21.0"
 spacy = "^3.4.1"
 sentencepiece = "^0.1.96"
```
Comment (saattrupdan) [inline pyproject.toml:15]:
```suggestion
sentencepiece = ">=0.1.96,<1.0.0"
```

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Looks great! Only a tiny change left in the dependencies.

**Review by saattrupdan — APPROVED** [review]
LGTM!
